ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,536 @@
+                    x,
+                )
+                .clone(),
+            });
+        });
+
+        // output
+        receive_right
+    }
+
+    /// This function generates a batch of quadratic proofs
+    /// it iteratively generates `ZeroKnowledgeProofs` and collects them in a batch
+    ///
+    /// It generates `prover_left_batch` as its left output
+    /// and `verifier_left_batch` as its right output.
+    /// Further, it outputs the generated masks included in the last proof of the batch.
+    ///
+    /// `PRSS` is invoked since `PRSS` right is used to generate `prover_left_batch`.
+    /// Since `PRSS` cannot be invoked later again on the same `Gate` and `Index`,
+    /// we use `PRSS` left to generate `verifier_left_batch` and output it as well.
+    ///
+    /// Outputs `(prover_left_batch, verifier_left_batch, (p_mask, q_mask))`
+    fn compute_proof_batch<C, I>(ctx: &C, uv_tuple_block: I) -> (Self, Self, (F, F))
+    where
+        C: Context,
+        I: Iterator<Item = UVTupleBlock<F>> + Clone,
+        R: ArrayLength + Sub<U1>,
+        <R as Sub<U1>>::Output: ArrayLength,
+    {
+        let mut prover_left_batch = Vec::<ZeroKnowledgeProof<F, R>>::new();
+        let mut verifier_left_batch = Vec::<ZeroKnowledgeProof<F, R>>::new();
+
+        // precomputation
+        // N is the proof length,
+        // i.e. polynomial of degree 2*R-1
+        // where R is the recursion factor
+        // the denominator takes as parameter R
+        let denominator = CanonicalLagrangeDenominator::<F, R>::new();
+        // the lagrange table takes as parameter R, and R-1
+        let lagrange_table = LagrangeTable::<F, R, <R as Sub<U1>>::Output>::from(denominator);
+
+        // set up record counter
+        let mut record_counter = 0usize;
+
+        // generate first proof from input
+        let mut uv_store = Self::compute_next_proof(
+            &mut prover_left_batch,
+            &mut verifier_left_batch,
+            &mut record_counter,
+            ctx,
+            &lagrange_table,
+            Self::polynomials_from_blocks::<R, _>(uv_tuple_block),
+        );
+
+        // recursively generate proofs
+        // until there are less than `R::USIZE * (R::USIZE-1)` many (u,v) field elements
+        // and then do one more iteration
+        // This ensures that after the last iteration,
+        // there are at most R::USIZE-1 many real (u,v) field elements
+        // Therefore at least the last field element is a `F:ZERO` filler that can be replaced
+        // Therefore we have space to include the masks during the final proof
+        //
+        // notice that uv_store.len() will always be at least 1
+        // further, polynomial uv_store.uv[0].0 and .1 will always have
+        // R::USIZE many points since they are filled with `F::ZERO`.
+        loop {
+            let stop = uv_store.len() < R::USIZE * (R::USIZE - 1);
```

Comment:
Since they are all constants, isn't that anyway precomputed by the compiler? 

R = 0 does not make sense, R = 1 means that there is 0 compression and by the definition of the zeroknowledge proofs it is supposed to be an infinite loop. I have added an assert that R is larger than 1.